src/app/codelabs/all-exercises/all-exercises.component.ts

```diff
@@ -0,0 +1,51 @@
+import { ExerciseBase } from './../../exercise/exercise/exercise.base';
+import { Component } from '@angular/core';
+import { Ng2TsExercises, ng2tsConfig } from '../../../../ng2ts/ng2ts';
+
+@Component({
+  selector: 'slides-all-exercises',
+  templateUrl: './all-exercises.component.html',
+  styleUrls: ['./all-exercises.component.css']
+})
+export class AllExercisesComponent {
+  private exercise;
+  private milestones;
+  public allFiles = {};
+
+  constructor(private exercises: Ng2TsExercises) {
+    this.exercise = exercises.getExercises(1, 1);
+    this.milestones = ng2tsConfig.milestones.map((milestone) => {
+      return {
+        name: milestone.name,
+        exercises: milestone.exercises
+          .map((exercise, index) => { return { ...exercise, index }; })
+          .filter((exercise) => {
+            if (exercise['files']) { return true; }
+          })
+      };
+    });
+
+    this.milestones.forEach((milestone) => {
+      milestone.exercises.forEach((exerciseFiles) => {
+        exerciseFiles.files.forEach((file) => {
+          if (this.allFiles[file.path]) {
+            this.allFiles[file.path].files.push(file);
+            this.allFiles[file.path].milestones.push(milestone.name);
+          } else {
+            this.allFiles[file.path] = {
+              files: [file],
+              milestones: [milestone.name],
+              name: file.path
+            };
+          }
+        });
+      });
+    });
+
+    this.allFiles = Object.keys(this.allFiles).map((file) => this.allFiles[file]);
```

Comment:
You know where you can do that? In Vue...

lol. 🥔

Anyway, check these out: [1](https://webcake.co/looping-over-maps-and-sets-in-angular-2s-ngfor/) [2](https://stackoverflow.com/questions/35534959/access-key-and-value-of-object-using-ngfor) [3](https://netbasal.com/angular-2-ngfor-array-with-unique-values-6b15478d6484) [4](https://webcake.co/looping-over-maps-and-sets-in-angular-2s-ngfor/)

It seems like you can't and won't ever be able to do this with Maps or Objects (until they maybe introduce some kind of pipe)

You can with Set, but I can't think of a reason why it would be better in this case, and seems like it might actually [be slower](https://stackoverflow.com/questions/39007637/javascript-set-vs-array-performance) 🤔